File: studentassistant/src/main/java/ru/erdenian/studentassistant/adapter/SchedulePagerAdapter.java
package ru.erdenian.studentassistant.adapter;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;

import org.joda.time.Days;
import org.joda.time.LocalDate;

import lombok.NonNull;
import ru.erdenian.studentassistant.fragment.SchedulePageFragment;
import ru.erdenian.studentassistant.schedule.ScheduleManager;
import ru.erdenian.studentassistant.schedule.Semester;

/**
 * Todo: описание класса.
 *
 * @author Ilya Solovyev
 * @version 0.0.0
 * @since 0.0.0
 */
public class SchedulePagerAdapter extends FragmentStatePagerAdapter {

    private final int COUNT;

    private final static String TITLE_FORMAT = "EEEE, dd MMMM";
    private final static String TITLE_FORMAT_FULL = "EEEE, dd MMMM yyyy";

    private Semester semester;

    public SchedulePagerAdapter(FragmentManager fm, @NonNull Semester semester) {
        super(fm);

        this.semester = semester;
        this.COUNT = semester.getLength();
    }

    @Override
    public CharSequence getPageTitle(int position) {
        LocalDate today = LocalDate.now();
        LocalDate day = semester.getFirstDay().plusDays(position);

        StringBuffer title = new StringBuffer();

        if (day.getYear() == today.getYear()) {
            title.append(day.toString(TITLE_FORMAT));
        } else {
            title.append(day.toString(TITLE_FORMAT_FULL));
        }

        return title;
    }

    @Override
    public Fragment getItem(int position) {
        return SchedulePageFragment.newInstance(ScheduleManager.getSemesters().asList().indexOf(semester),
                semester.getFirstDay().plusDays(position));
    }

    @Override
    public int getCount() {
        return COUNT;
    }

    public int getPosition(@NonNull LocalDate date) {
        return Days.daysBetween(semester.getFirstDay(), date).getDays();
    }
}

File: studentassistant/src/main/kotlin/ru/erdenian/studentassistant/adapter/SchedulePagerAdapter.kt
package ru.erdenian.studentassistant.adapter

import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentStatePagerAdapter
import org.joda.time.Days
import org.joda.time.LocalDate
import ru.erdenian.studentassistant.fragment.SchedulePageFragment
import ru.erdenian.studentassistant.schedule.Semester

class SchedulePagerAdapter(fm: FragmentManager, private val semester: Semester, val showWeeksAndDates: Boolean) :
    FragmentStatePagerAdapter(fm) {

  companion object {
    private const val TITLE_FORMAT = "EEEE, dd MMMM"
    private const val TITLE_FORMAT_FULL = "EEEE, dd MMMM yyyy"

    private val weekdays = Array<String>(7) { LocalDate().withDayOfWeek(it + 1).dayOfWeek().asText }
  }

  override fun getPageTitle(position: Int): CharSequence =
      if (showWeeksAndDates) weekdays[position]
      else semester.firstDay.plusDays(position).run {
        if (year == LocalDate.now().year) toString(TITLE_FORMAT)
        else toString(TITLE_FORMAT_FULL)
      }

  override fun getItem(position: Int): Fragment =
      if (showWeeksAndDates) SchedulePageFragment.newInstance(semester.id, position + 1)
      else SchedulePageFragment.newInstance(semester.id, getDate(position))

  override fun getCount(): Int = if (showWeeksAndDates) 7 else semester.length

  fun getPosition(date: LocalDate): Int =
      if (showWeeksAndDates) throw UnsupportedOperationException("showWeeksAndDates = $showWeeksAndDates")
      else Days.daysBetween(semester.firstDay, date).days

  fun getDate(position: Int): LocalDate =
      if (showWeeksAndDates) throw UnsupportedOperationException("showWeeksAndDates = $showWeeksAndDates")
      else semester.firstDay.plusDays(position)
}
